### ISSUE TITLE
MacOS Desktop:  Changing caps Lock key input source throws exception

### ISSUE THREAD

**Issue by PassionPenguin** [flutter/flutter#67866]
**What I did: Preference > Keyboard > Use the Caps Lock key to switch to and from ABC
What it caused: Only Console Log**

════════ Exception caught by services library ════════ 
Attempted to send a key down event when no keys are in keysPressed. This state can occur if the key event being sent doesn't properly set its modifier flags. This was the event: RawKeyDownEvent#ef1aa(logicalKey: LogicalKeyboardKey#70039(keyId: "0x100070039", keyLabel: "", debugName: "Caps Lock"), physicalKey: PhysicalKeyboardKey#70039(usbHidUsage: "0x00070039", debugName: "Caps Lock")) and its data: RawKeyEventDataMacOs(keyLabel: , keyCode: 57, characters: , unmodifiedCharacters: , modifiers: 256, modifiers down: {})
'package:flutter/src/services/raw_keyboard.dart':
Failed assertion: line 601 pos 12: '<optimized out>'
════════ ════════ ════════ ════════ ════════

**Comment by PassionPenguin** [comment]
<img width="780" alt="Screenshot 2020-10-11 at 11 25 43 PM" src="https://user-images.githubusercontent.com/41184947/95682663-17538f80-0c19-11eb-8ac1-38cae03a65b4.png">

The Second Option of the **Keyboard>Input Sources** Preference in **_System Preferences.app_**.

**Comment by PassionPenguin** [comment]
Related Code: [raw_keyboard.dart:617](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/services/raw_keyboard.dart#L617)
```
    assert(event is! RawKeyDownEvent || _keysPressed.isNotEmpty,
        'Attempted to send a key down event when no keys are in keysPressed. '
        "This state can occur if the key event being sent doesn't properly "
        'set its modifier flags. This was the event: $event and its data: '
        '${event.data}');```

**Comment by darshankawar** [comment]
@PassionPenguin 
Please provide `flutter doctor -v` and a minimal complete reproducible code sample and detailed steps to replicate. Also, is this on MacOS desktop ?

**Comment by darshankawar** [comment]
Similar to [63939](https://github.com/flutter/flutter/issues/63939) where the author mentioned `no problem with macos`

**Comment by PassionPenguin** [comment]
> @PassionPenguin
> Please provide `flutter doctor -v` and a minimal complete reproducible code sample and detailed steps to replicate. Also, is this on MacOS desktop ?

In fact any code with [caps lock using for language changing] can cause the problem.
In addition, closing other applications with Cmd+Q can also causes this problem.

**Comment by darshankawar** [comment]
@PassionPenguin 
So when does the error occur ? If you change the input sources from `caps lock off` to `caps lock on`, what steps do we need to follow to see the error ?

**Comment by PassionPenguin** [comment]
actually the `caps lock` key is still off when press if this feature enabled, and only long press will make caps lock on.
just use a simple input textfield is enough to reproduce the error log

**Comment by darshankawar** [comment]
Ok, so I was able to replicate the issue. Steps I took to replicate.

1. Create fresh project (have macOS desktop configuration enabled)
2. Add a `TextField`.
3. Run app.
4. Go to `System Preferences -> Keyboard -> Input Sources -> Tab Caps Lock key to change input source (caps lock OFF to ON).
5. Come back to app. Input anything in textfield and press caps lock, which throws below exception:

```
An Observatory debugger and profiler on macOS is available at: http://127.0.0.1:49213/eX_7F7tzjKU=/
flutter: ══╡ EXCEPTION CAUGHT BY SERVICES LIBRARY ╞══════════════════════════════════════════════════════════
flutter: The following assertion was thrown during a platform message callback:
flutter: Attempted to send a key down event when no keys are in keysPressed. This state can occur if the key
flutter: event being sent doesn't properly set its modifier flags. This was the event:
flutter: RawKeyDownEvent#f5c73(logicalKey: LogicalKeyboardKey#70039(keyId: "0x100070039", keyLabel: "",
flutter: debugName: "Caps Lock"), physicalKey: PhysicalKeyboardKey#70039(usbHidUsage: "0x00070039",
flutter: debugName: "Caps Lock")) and its data: RawKeyEventDataMacOs(keyLabel: , keyCode: 57, characters: ,
flutter: unmodifiedCharacters: , modifiers: 256, modifiers down: {})
flutter: 'package:flutter/src/services/raw_keyboard.dart':
flutter: Failed assertion: line 618 pos 12: '<optimized out>'
flutter: 
flutter: Either the assertion indicates an error in the framework itself, or we should provide substantially
flutter: more information in this error message to help you determine and fix the underlying cause.
flutter: In either case, please report this assertion by filing a bug on GitHub:
flutter:   https://github.com/flutter/flutter/issues/new?template=BUG.md
flutter: 
flutter: When the exception was thrown, this was the stack:
flutter: #2      RawKeyboard._handleKeyEvent (package:flutter/src/services/raw_keyboard.dart:618:12)
flutter: #3      BasicMessageChannel.setMessageHandler.<anonymous closure> (package:flutter/src/services/platform_channel.dart:74:49)
flutter: #4      _DefaultBinaryMessenger.handlePlatformMessage (package:flutter/src/services/binding.dart:283:33)
flutter: #5      _invoke3.<anonymous closure> (dart:ui/hooks.dart:280:15)
flutter: #9      _invoke3 (dart:ui/hooks.dart:279:10)
flutter: #10     _dispatchPlatformMessage (dart:ui/hooks.dart:154:5)
flutter: (elided 5 frames from class _AssertionError and dart:async)
flutter: ════════════════════════════════════════════════════════════════════════════════════════════════════

```

**Comment by PassionPenguin** [comment]
in addition, Command+Q can also produce a error as i mentioned above:)

just focus the cursor into the flutter textfield, then use this hot key to close another app, and it will show:

════════ Exception caught by services library ══════════════════════════════════════════════════════
Attempted to send a key down event when no keys are in keysPressed. This state can occur if the key event being sent doesn't properly set its modifier flags. This was the event: RawKeyDownEvent#bc55d(logicalKey: LogicalKeyboardKey#700e3(keyId: "0x1000700e3", keyLabel: "", debugName: "Meta Left"), physicalKey: PhysicalKeyboardKey#700e3(usbHidUsage: "0x000700e3", debugName: "Meta Left")) and its data: RawKeyEventDataMacOs(keyLabel: , keyCode: 55, characters: , unmodifiedCharacters: , modifiers: 256, modifiers down: {})
'package:flutter/src/services/raw_keyboard.dart':
Failed assertion: line 601 pos 12: '<optimized out>'
════════════════════════════════════════════════════════════════════════════════════════════════════

**Comment by PassionPenguin** [comment]
maybe the solution is to add another two condition statement and to ignore the event :)

**Comment by PassionPenguin** [comment]
and cupertinotextfield can cause the same result...

**Comment by PassionPenguin** [comment]
#74138 `Shift` key

**Comment by darshankawar** [comment]
I am unable to replicate this error using latest versions and using steps to replicate mentioned [here](https://github.com/flutter/flutter/issues/67866#issuecomment-710010582). The app doesn't crash and there are no crash logs in the console.

```
dhs@Dhss-MacBook-Pro newtest % flutter run -d macos
Launching lib/main.dart on macOS in debug mode...
/Users/dhs/Documents/NCFlutter/newtest/macos/Pods/Pods.xcodeproj: warning: The macOS deployment target 'MACOSX_DEPLOYMENT_TARGET' is set to 10.7, but the range of supported deployment target versions is 10.9 to 12.2. (in target 'leveldb-library' from project 'Pods')
Building macOS application...                                           
Syncing files to device macOS...                                   140ms

Flutter run key commands.
r Hot reload. 🔥🔥🔥
R Hot restart.
h List all available interactive commands.
d Detach (terminate "flutter run" but leave application running).
c Clear the screen
q Quit (terminate the application on the device).

💪 Running with sound null safety 💪

An Observatory debugger and profiler on macOS is available at: http://127.0.0.1:53958/258_K9scSeE=/
The Flutter DevTools debugger and profiler on macOS is available at: http://127.0.0.1:9100?uri=http://127.0.0.1:53958/258_K9scSeE=/


```

Closing this as fixed.

<details>
<summary> stable, master flutter doctor -v </summary>

```
[✓] Flutter (Channel stable, 3.3.8, on macOS 12.2.1 21D62 darwin-x64, locale
    en-GB)
    • Flutter version 3.3.8 on channel stable at
      /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 52b3dc25f6 (25 hours ago), 2022-11-09 12:09:26 +0800
    • Engine revision 857bd6b74c
    • Dart version 2.18.4
    • DevTools version 2.15.0

[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.

[!] Flutter (Channel master, 3.6.0-7.0.pre.18, on macOS 12.2.1 21D62 darwin-x64,
    locale en-GB)
    • Flutter version 3.6.0-7.0.pre.18 on channel master at
      /Users/dhs/documents/fluttersdk/flutter
    ! Warning: `flutter` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/flutter, which is not inside
      your current Flutter SDK checkout at
      /Users/dhs/documents/fluttersdk/flutter. Consider adding
      /Users/dhs/documents/fluttersdk/flutter/bin to the front of your path.
    ! Warning: `dart` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/dart, which is not inside your
      current Flutter SDK checkout at /Users/dhs/documents/fluttersdk/flutter.
      Consider adding /Users/dhs/documents/fluttersdk/flutter/bin to the front
      of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 2921ca0c48 (2 hours ago), 2022-11-21 21:43:26 -0500
    • Engine revision 8dd8e092e4
    • Dart version 2.19.0 (build 2.19.0-415.0.dev)
    • DevTools version 2.19.0
    • If those were intentional, you can disregard the above warnings; however
      it is recommended to use "git" directly to perform update checks and
      upgrades.

[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.



```

</details>
